tuf/repository/_repository.py

```diff
@@ -0,0 +1,139 @@
+        """Context manager for editing a roles metadata
+
+        Context manager takes care of loading the roles metadata (or creating
+        new metadata if 'init'), updating expiry and version. The caller can do
+        other changes to the Signed object and when the context manager exits,
+        a new version of the roles metadata is stored.
+
+        Context manager user can raise AbortEdit from inside the with-block to
+        cancel the edit: in this case none of the changes are stored.
+        """
+        md = self.open(role, init)
+        with suppress(AbortEdit):
+            yield md.signed
+            self.close(role, md)
+
+    def sign(self, role: str) -> None:
+        """sign without modifying content, or removing existing signatures"""
+        md = self.open(role)
+        self.close(role, md, sign_only=True)
+
+    def snapshot(
+        self, current_targets: Dict[str, MetaFile]
+    ) -> Tuple[Optional[int], Dict[str, MetaFile]]:
+        """Update snapshot meta information
+
+        Updates the meta information in snapshot according to input.
+
+        Arguments:
+            current_targets: The new currently served targets roles.
+
+        Returns: Tuple of
+            - New snapshot version or None if snapshot was not created
+            - Meta information for targets metadata that were removed from repository
+        """
+
+        # Snapshot update is needed if
+        # * any targets files are not yet in snapshot or
+        # * any targets version is incorrect
+        updated_snapshot = False
+        removed: Dict[str, MetaFile] = {}
+
+        with self.edit("snapshot") as snapshot:
+            for keyname, new_meta in current_targets.items():
+                if keyname not in snapshot.meta:
+                    updated_snapshot = True
+                    snapshot.meta[keyname] = new_meta
+                    continue
+
+                old_meta = snapshot.meta[keyname]
+                if new_meta.version < old_meta.version:
+                    raise ValueError(f"{keyname} version rollback")
+                if new_meta.version > old_meta.version:
+                    updated_snapshot = True
+                    snapshot.meta[keyname] = new_meta
+                    removed[keyname] = old_meta
+
+            if not updated_snapshot:
+                # prevent edit() from storing a new snapshot version
+                raise AbortEdit("Skip snapshot: No targets version changes")
+
+        if not updated_snapshot:
+            # This code is reacheable as edit() handles AbortEdit
+            logger.debug("Snapshot update not needed")  # type: ignore[unreachable]
+        else:
+            logger.debug(
+                "Snapshot v%d, %d targets", snapshot.version, len(snapshot.meta)
+            )
+
+        version = snapshot.version if updated_snapshot else None
+        return version, removed
+
+    def timestamp(self, snapshot_meta: MetaFile) -> Optional[MetaFile]:
```

Comment:
one thing I don't like WRT snapshot/timestamp here (in addition to arguments mentioned in another comment):

timestamp() always creates a new version, snapshot does not.
* this is practically useful as timestamp does not get called "just in case" but snapshot might (a repository might not know if snapshot is needed or not: are there new targets or not?)
* on the other hand especially timestamp should get called ""without reason", just to bump the timestamp
* this is annoyingly unbalanced... so maybe should add a `force` argument to both to allow either mode -- "always create new version" or "only create if needed"

Reply:
modified this, resolving